src/Adapter/MSTest.TestAdapter/Discovery/AssemblyEnumerator.cs

```diff
@@ -278,13 +297,82 @@ private bool DynamicDataAttached(IDictionary<string, object?> sourceLevelParamet
-        using var writer = new StringWriter();
-        TestMethod testMethod = test.TestMethod;
-        MSTestAdapter.PlatformServices.Interface.ITestContext testContext = PlatformServiceProvider.Instance.GetTestContext(testMethod, writer, sourceLevelParameters);
-        TestMethodInfo? testMethodInfo = _typeCache.GetTestMethodInfo(testMethod, testContext, MSTestSettings.CurrentSettings.CaptureDebugTraces);
-        return testMethodInfo != null && TryProcessTestDataSourceTests(test, testMethodInfo, tests);
+        return testMethodInfo.Value != null && TryProcessTestDataSourceTests(test, testMethodInfo.Value, tests);
+    }
+
+    private static void AddFixtureTests(TestMethodInfo testMethodInfo, List<UnitTestElement> tests, HashSet<string> fixtureTests)
+    {
+        string assemblyName = testMethodInfo.Parent.Parent.Assembly.GetName().Name!;
+        string assemblyLocation = testMethodInfo.Parent.Parent.Assembly.Location;
+        string className = testMethodInfo.Parent.ClassType.Name;
+        string classFullName = testMethodInfo.Parent.ClassType.FullName!;
+
+        if (!fixtureTests.Contains(assemblyLocation))
+        {
+            _ = fixtureTests.Add(assemblyLocation);
+
+            if (testMethodInfo.Parent.Parent.AssemblyInitializeMethod is not null)
+            {
+                tests.Add(GetAssemblyFixtureTest(testMethodInfo.Parent.Parent.AssemblyInitializeMethod, assemblyName, className, classFullName, assemblyLocation, Constants.AssemblyInitializeFixtureTrait));
+            }
+
+            if (testMethodInfo.Parent.Parent.AssemblyCleanupMethod is not null)
+            {
+                tests.Add(GetAssemblyFixtureTest(testMethodInfo.Parent.Parent.AssemblyCleanupMethod, assemblyName, className, classFullName, assemblyLocation, Constants.AssemblyCleanupFixtureTrait));
+            }
+        }
+
+        if (!fixtureTests.Contains(assemblyLocation + classFullName))
+        {
+            _ = fixtureTests.Add(assemblyLocation + classFullName);
+
+            if (testMethodInfo.Parent.ClassInitializeMethod is not null)
+            {
+                tests.Add(GetClassFixtureTest(testMethodInfo.Parent.ClassInitializeMethod, assemblyName, className, classFullName, assemblyLocation, Constants.ClassInitializeFixtureTrait));
+            }
+
+            if (testMethodInfo.Parent.ClassCleanupMethod is not null)
+            {
+                tests.Add(GetClassFixtureTest(testMethodInfo.Parent.ClassCleanupMethod, assemblyName, className, classFullName, assemblyLocation, Constants.ClassCleanupFixtureTrait));
+            }
+        }
+
+        static UnitTestElement GetAssemblyFixtureTest(MethodInfo methodInfo, string assemblyName, string className, string classFullName,
+            string assemblyLocation, string fixtureType)
+        {
+            string methodName = GetMethodName(methodInfo);
+            string[] hierarchy = [null!, assemblyName, $"[{Constants.AssemblyFixturesHierarchyName}]", methodName];
```

Comment:
Do we need the parameters? cc @drognanar

Reply:
Yes. These parameters form the tree node